src/pipert2/core/base/pipe.py

```diff
@@ -1,2 +1,93 @@
+from src.pipert2.core.managers.network import Network
+
+
 class Pipe:
-    pass
+    """The pipe object is the center of the pipe.
+    Once it is created it act as a central registry for the
+    pipe methods such as create flows, register routines, notify events
+    and more.
+
+    """
+
+    def __init__(self, networking: Network, logger: PipeLogger):  # TODO - default logger and default networking (Queue)
+        """
+        Args:
+            networking: Network object responsible for the routine's communication.
+            logger: PipeLogger object for logging the pipe actions.
+
+        Attributes:
+            networking: Network object responsible for the routine's communication.
+            logger: PipeLogger object for logging the pipe actions.
+            flows (dict[str, Flow]): Dictionary mapping the pipe flows to their name.
+            event_board (EventBoard): EventBoard object responsible for the pipe events.
+
+        """
+
+        self.network = networking
+        self.logger = logger
+        self.flows = {}
+        self.event_board = EventBoard()
+
+    def create_flow(self, flow_name: str, auto_wire: bool, *routines: Routine):
+        """Create a new flow in the pipe.
+
+        Args:
+            flow_name (str): The name of the flow to be created
+            auto_wire (bool): Automatically connect the routines to each other by the order of their entry.
+            routines: (Routine): List of routines to register to the flow.
+
+        """
+
+        for routine in routines:
+            routine.initialize(message_handler=self.network.get_message_handler(routine.name),
+                               event_notifier=self.event_board.get_event_notifier())
+
+        flow = Flow(flow_name, self.event_board, self.logger.get_child(), routines=list(routines))
+        self.flows[flow_name] = flow
+
+        if auto_wire:
+            for first_routine, second_routine in zip(routines, routines[1:]):
+                self.network.link(src=first_routine, destinations=second_routine)
+
+    def link(self, *wires):
+        """Connect the routines to each other by their wires configuration.
+
+        Args:
+            wires (Wire): List of wires to connect their routines
+
+        """
+
+        for wire in wires:
+            self.network.link(src=wire.source, destinations=wire.destinations)
+
+    def build(self):
+        """Build the pipe to be ready to start working.
+
+        """
+
+        for flow in self.flows.values():
+            flow.build()
+
+    def notify_event(self, event_name: str, **event_parameters) -> None:
+        """Notify an event has started
+
+        Args:
+            event_name: The name of the event to notify
+
+        """
+
+        self.event_board.notify_event(event_name, **event_parameters)
+
+    def join(self):
+        """Block the execution until all of the flows have been killed
+
+        """
+
+        # TODO - Maybe before joining send a kill event ?
```

Comment:
why to send a kill event?

Reply:
It depends,
If we know that the user only using the pipe in a single main thread then after joining he will wait forever,
this is happening because all of the flows processes will stay alive and there won't be any way to stop them :see_no_evil:

But maybe he does it on purpose ?

Actually after writing this i think i will add a parameter if that the user can specify whether he wants to send a kill event or not :sweat_smile: